### ISSUE TITLE
JSON parsing does not work

### ISSUE THREAD

**Issue by aurimasmb** [larsgw/citation.js#58]
`cite.set()` fails to parse standard JSON inputs, and object inputs.

I've reproduced this locally, and in the browser demo.

If you copy the JSON output produced in the browser demo back into one of the input windows, you get the following:

` [{"id":"temp_id_9016360104836194"}]`

Seems like `cite.set()` is failing to parse?

I'm trying to do something like:

``` js

var Cite = require('citation-js')

cite = new Cite();

var optionsObject = {
   format: 'object'
}

var optionsHTMLString = {
  format: 'string',
  type: 'html'
}

var inputData = "a list of references in conventional .bib format";

function displayFilteredAndSortedReferences() {
    
    // This works fine.
    cite.set(inputData);
    var referencesObj = cite.get(optionsObject);

    // Do some filtering / sorting of referencesObj (which requires them to be in object format.)

    // And now only output the filtered and sorted array of references to HTML to the page.

    // This does not work.
    cite.set(filteredReferencesObj);
    var outputHTML = cite.get(optionsHTML);

    //  Result:
    //  <div class="csl-bib-body">
    //  <div data-csl-entry-id="temp_id_6629628389089841" class="csl-entry"> (n.d.).</div> 
    //  </div>

    // This also doesn't work.
    cite.set(JSON.stringify(filteredReferencesObj));
    var outputHTML = cite.get(optionsHTML);
   
    //   Result:
    //  </div>
    //  <div class="csl-bib-body">
    //  </div>



}

```

**Comment by larsgw** [comment]
Not sure what is happening. Can you give a minimal example of failing input for `filteredReferencesObj`, so I can reproduce/debug myself?

<strike>One problem might be that the output options are incorrect: `'object'` is an incorrect value for the `format` option.</strike> I see you're using `optionsHTML` as options, but I don't see them defined. What are they?

**Comment by aurimasmb** [comment]
Hi Lars, after putting together the minimal code sample, I think the error was actually in my filtering code. However, it was producing an error which is similar to one which I was able to reproduce in the browser demo:

When you input a JSON string into the input boxes, it fails to parse.

`[{"id":"Q23571040","type":"article-journal","title":"Correlation of the Base Strengths of Amines 1","DOI":"10.1021/ja01577a030","author":[{"given":"H. K.","family":"Hall"}],"issued":{"date-parts":[[1957,1,1]]},"container-title":"Journal of the American Chemical Society","volume":"79","issue":"20","page":"5441-5444"}]`

![input](https://user-images.githubusercontent.com/1960526/29531109-f8a0e626-869e-11e7-8a9f-4cb7ebfd2785.png)

![output](https://user-images.githubusercontent.com/1960526/29531113-fb67c820-869e-11e7-8489-6d7e45aca92e.png)

Changing the `format` option to `real` fixed parsing for me. It now correctly parses into an object. Maybe it would be helpful to throw an error message for the `format = 'object'` condition. As of now it fails silently and outputs a JSON string.

Here's the working minimal example:

``` js

var fs = require('fs');
var Cite = require('citation-js');


var filePath = './example.bib';

var cite = new Cite();

var optionsObjectOutput = {
	format: 'real' // I was setting this to 'object' for some reason. Should have read the Docs more closely.  https://github.com/larsgw/citation.js#cite.in.options
	// Issue: setting this to 'object' does not throw an error. It just defaults to outputting as a string.
}

var optionsHTMLOutput = {
	format: 'string',
	type: 'html'
}

displayFilteredAndSortedReferences(filePath);

function displayFilteredAndSortedReferences(path) {
	fs.readFile(path, 'utf8', function(err, data) { 

		// Parse in .bib file data.
		cite.set(data);
		var referenceLibraryObjects = cite.get(optionsObjectOutput); 

		//console.log("Before Filtering: ------------------------------------------ ");
		//console.log(referenceLibraryObjects);

		// Filter reference objects.
		var filteredReferenceLibraryObjects = referenceLibraryObjects.filter(function(reference){
			return reference.id == 'allison_managing_2013' || reference.id == 'lee_stress-constrained_2012'; // Just extracting two example references.
		});

		//console.log("After Filtering: ------------------------------------------ ");
		//console.log(filteredReferenceLibraryObjects);

		// Output HTML display of citation.
		var referencesOutputHTML = [];
		filteredReferenceLibraryObjects.forEach(function(referenceObject) {
			cite.set(referenceObject);
			referencesOutputHTML.push(cite.get(optionsHTMLOutput));
		});

		// Works!
		console.log("HTML Output: ------------------------------------------ ");
		console.log(referencesOutputHTML);

	});
}

```

**Comment by larsgw** [comment]
1. I put the invalid-options-don't-throw-errors thing in a separate issue #59 
2. I am able to reproduce the issue
3. After playing around with the internal general parsing methods (`Cite.parse.input`), I found that the input type recognising function thinks the input is a [Bib.TXT](http://bibtxt.github.io/) string
4. You can expect a fix soon, in the next version (`v0.3.0-13`), today or tomorrow
